pkg/scheduler/queue/multi_queuing_algorithm_tree_queue_test.go

```diff
@@ -0,0 +1,982 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package queue
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+// enqueueObj is intended for use in tests; it represents an obj to enqueue to path, or an expected path and obj
+// for a given dequeue
+type enqueueObj struct {
+	obj  any
+	path QueuePath
+}
+
+func newTenantQuerierAssignments() *tenantQuerierAssignments {
+	return &tenantQuerierAssignments{
+		tenantQuerierIDs: make(map[TenantID]map[QuerierID]struct{}),
+		tenantNodes:      make(map[string][]*Node),
+	}
+}
+
+func Test_NewTree(t *testing.T) {
+	tests := []struct {
+		name      string
+		treeAlgos []QueuingAlgorithm
+		state     *tenantQuerierAssignments
+		expectErr bool
+	}{
+		{
+			name:      "create round-robin tree",
+			treeAlgos: []QueuingAlgorithm{&roundRobinState{}},
+		},
+		{
+			name:      "create tenant-querier tree",
+			treeAlgos: []QueuingAlgorithm{&tenantQuerierAssignments{}},
+		},
+		{
+			name:      "fail to create tree without defined dequeuing algorithm",
+			treeAlgos: []QueuingAlgorithm{nil},
+			expectErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := NewTree(tt.treeAlgos...)
+			if tt.expectErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
+
+		})
+	}
+}
+
+func Test_EnqueueBackByPath(t *testing.T) {
```

Comment:
lots of test cases here that are just like "there was no error", it's not clear to me what is being tested - do we want to verify the nodes were created as expected or something else?

Reply:
We could additionally verify that dequeuing removes all the same enqueued objects in order; I was split on whether to do so, since we also have tests on dequeue behavior.

Reply:
maybe just we do something like a `getNode` and ensure its item count increased after enqueue?

The dequeue-after-enqueue tests do cover a lot but they also depend on the dequeuing behavior, it would be nice to have something specifically just verifying how the enqueue works in isolation

Reply:
How would you feel about checking that `ItemCount()` matches what we expect? I've said before, I'd like to avoid internally peeping tree structure in tests; `ItemCount()` is something we already use/require as part of the tree interface, since we need it to limit enqueued requests.

Reply:
> checking that ItemCount() on the whole tree matches what we expect

I've done this